ref-farming/src/actions_of_reward.rs

```diff
@@ -153,6 +161,59 @@ fn claim_user_reward_from_farm(
 }
 
 impl Contract {
+    
+    /// if token was given, then just withdraw that token, or withdraw all tokens
+    pub(crate) fn internal_withdraw_all_rewards(&mut self, farmer_id: &AccountId, token: Option<&AccountId>) {
+        
+        let view_farmer = self.get_farmer(&farmer_id);
+        let tokens = {
+            if let Some(token) = token {
+                vec![token]
+            } else {
+                view_farmer.get_ref().rewards.keys().collect::<Vec<_>>()
+            }
+        };
+        
+        let mut farmer = self.get_farmer(&farmer_id);
+        let mut withdraw_tasks = HashMap::<AccountId, u128>::new();
+        for token_id in tokens {
+            let amount = farmer.get_ref_mut().sub_reward(&token_id, 0);
+            withdraw_tasks.insert(token_id.clone(), amount);
+        }
+        self.data_mut().farmers.insert(farmer_id, &farmer);
+
+        for (token_id, amount) in withdraw_tasks {
```

Comment:
Hashmap is expensive to construct and unnecessary here. A `Vec` should work as well.
```suggestion
        let withdraw_amounts: Vec<_> = tokens.iter().map(|token_id| farmer.get_ref_mut().sub_reward(&token_id, 0)).collect();
        self.data_mut().farmers.insert(farmer_id, &farmer);
        
        for (token_id, amount) in tokens.into_iter().zip(withdraw_amounts) {
```